src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs

```diff
@@ -580,14 +626,260 @@ static int GetRawKdfCount(in AlgorithmIdentifierAsn algorithmIdentifier)
+            // This all breaks down to:
+            //
+            // if (machine key && can't make machine keys && known CAPI provider)
+            // {
+            //     if (RSA)
+            //     {
+            //         return CAPI PROV_RSA_AES;
+            //     }
+            //     else
+            //     {
+            //         pretend PreserveStorageProvider was set.
+            //     }
+            // }
+            //
+            // return CNG KSP;
+
+            if (HasMachineKey(bagAttributes, storageFlags))
+            {
+                if (HasCapiCsp(bagAttributes, out bool isRsa))
+                {
+                    if (machineKeyState == CngMachineKeyState.Unknown)
+                    {
+                        machineKeyState = CheckMachineKeyPermissions();
+                    }
+
+                    if (machineKeyState == CngMachineKeyState.Denied)
+                    {
+                        if (isRsa)
+                        {
+                            return s_syntheticCapiCspAttribute;
+                        }
+
+                        // A DSS or DH CAPI provider, preserve it.
+                        return null;
+                    }
+                }
+            }
+
+            // Not a machine key, or wasn't known to be a CAPI key, so use CNG.
+            return s_syntheticKspAttribute;
+
+            static bool HasMachineKey(AttributeAsn[]? bagAttributes, X509KeyStorageFlags storageKind)
+            {
+                // This needs to be kept in sync with the Windows MapStorageFlags behavior.
+                // We say that both UserKeySet & MachineKeySet maps to a user key.
+
+                if ((storageKind & X509KeyStorageFlags.UserKeySet) != 0)
+                {
+                    return false;
+                }
+
+                if ((storageKind & X509KeyStorageFlags.MachineKeySet) != 0)
+                {
+                    return true;
+                }
+
+                // The Machine Key Set attribute OID is a sentinel, the attribute has no interpreted value.
+                // So, any attribute that specifies it means it's a machine key.
+                if (bagAttributes is not null)
+                {
+                    foreach (AttributeAsn attr in bagAttributes)
+                    {
+                        if (attr.AttrType == Oids.MsPkcs12MachineKeySet)
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            static CngMachineKeyState CheckMachineKeyPermissions()
+            {
+#if NET || NETFRAMEWORK
+                CngKey? cngKey = null;
+
+                try
+                {
+#if NET
+                    Debug.Assert(OperatingSystem.IsWindows());
 #endif
 
+                    cngKey = CngKey.Create(
+                        CngAlgorithm.Rsa,
+                        Guid.NewGuid().ToString("B"),
```

Comment:
Can we prefix this with something? That way if someone sees machine keys appearing and disappearing in the Windows Event Log they have a chance of at least just identifying what it could be.